pkg/ring/consul_client.go

```diff
@@ -203,6 +207,55 @@ func (c *consulClient) WatchKey(ctx context.Context, key string, f func(interfac
 	}
 }
 
+// WatchPrefix will watch a given prefix in Consul for new keys and changes to existing keys under that prefix.
+// When the value under said key changes, the f callback is called with the deserialised value.
+// To construct the deserialised value, a factory function should be supplied which generates an empty
+// struct for WatchKey to deserialise into. Note that for WatchPrefix it's a struct containing the Consul key and
+// value KVPair that is passed to the callback f. See the KVEntry struct in kvstore.go. Values in Consul are assumed
+// to be JSON. This function blocks until the context is cancelled.
+func (c *consulClient) WatchPrefix(ctx context.Context, prefix string, f func(interface{}) bool) {
+
+	var (
+		backoff = util.NewBackoff(ctx, backoffConfig)
+		index   = uint64(0)
+	)
+	for backoff.Ongoing() {
+		queryOptions := &consul.QueryOptions{
+			RequireConsistent: true,
+			WaitIndex:         index,
+			WaitTime:          longPollDuration,
+		}
+
+		kvps, meta, err := c.kv.List(prefix, queryOptions.WithContext(ctx))
+		if err != nil || kvps == nil {
+			level.Error(util.Logger).Log("msg", "error getting path", "prefix", prefix, "err", err)
+			backoff.Wait()
+			continue
+		}
+		backoff.Reset()
+		// Skip if the index is the same as last time, because the key value is
+		// guaranteed to be the same as last time
+		if index == meta.LastIndex {
+			continue
+		}
+
+		index = meta.LastIndex
+		for _, kvp := range kvps {
+			out, err := c.codec.Decode(kvp.Value)
+			if err != nil {
+				level.Error(util.Logger).Log("msg", "error decoding list of values for prefix:key", "prefix", prefix, "key", kvp.Key, "err", err)
+				continue
+			}
+			if !f(KVEntry{
+				Cluster:      kvp.Key,
+				ReplicaEntry: out,
+			}) {
```

Comment:
Can we pass `f` a key and value instead of reconstructing this?  Signature probably wants to be `func(string, interface{})`.